ec2_customer_gateway: improve returned data

s-hertel opened the pull request:
##### SUMMARY
Make the returned results concise by removing response metadata and camel_dict_to_snake_dict the results. Ensure customer_gateways is always returned as the documentation implied. Return the value of the Name tag instead of the metadata returned by tagging. This also allows results['name'] to be utilized when the customer_gateway exists. Added to the docs. Add tags to the results returned during customer gateway creation.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py

##### ANSIBLE VERSION
```
2.4.0
```

##### ADDITIONAL INFORMATION
Here is a sample task to see the difference of the returned values (to ensure the same path taken, each time before running it I ensured there was no customer gateway with that IP):
```
---
- hosts: localhost
  gather_facts: false
  connection: local
  tasks:
    - name: create customer gateway
      ec2_customer_gateway:
        profile: shertel
        routing: static
        ip_address: 2.3.4.5
        name: AnsibleOffice
        region: us-east-1
      register: cgw
```

Before fix:
```
changed: [localhost] => {
    "changed": true,
    "failed": false,
    "gateway": {
        "customer_gateway": {
            "bgp_asn": "65000",
            "customer_gateway_id": "cgw-3ff21956",
            "ip_address": "2.3.4.5",
            "state": "available",
            "type": "ipsec.1"
        },
        "response_metadata": {
            "http_headers": {
                "content-type": "text/xml;charset=UTF-8",
                "date": "Thu, 03 Aug 2017 14:45:16 GMT",
                "server": "AmazonEC2",
                "transfer-encoding": "chunked",
                "vary": "Accept-Encoding"
            },
            "http_status_code": 200,
            "request_id": "0df40d39-5533-483f-9467-7f129de91d9c",
            "retry_attempts": 0
        }
    },
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "bgp_asn": null,
            "ec2_url": null,
            "ip_address": "2.3.4.5",
            "name": "AnsibleOffice",
            "profile": "shertel",
            "region": "us-east-1",
            "routing": "static",
            "security_token": null,
            "state": "present",
            "validate_certs": true
        }
    },
    "name": {
        "response_metadata": {
            "http_headers": {
                "content-type": "text/xml;charset=UTF-8",
                "date": "Thu, 03 Aug 2017 14:45:16 GMT",
                "server": "AmazonEC2",
                "transfer-encoding": "chunked",
                "vary": "Accept-Encoding"
            },
            "http_status_code": 200,
            "request_id": "262c7094-32b3-4f60-a888-aba1821e97d1",
            "retry_attempts": 0
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```
After fix:
```
changed: [localhost] => {
    "changed": true,
    "failed": false,
    "gateway": {
        "customer_gateway": {
            "bgp_asn": "65000",
            "customer_gateway_id": "cgw-3bf21952",
            "ip_address": "2.3.4.5",
            "state": "available",
            "tags": {
                "Name": "AnsibleOffice"
            },
            "type": "ipsec.1"
        },
        "customer_gateways": [
            {
                "bgp_asn": "65000",
                "customer_gateway_id": "cgw-3bf21952",
                "ip_address": "2.3.4.5",
                "state": "available",
                "tags": {
                    "Name": "AnsibleOffice"
                },
                "type": "ipsec.1"
            }
        ]
    },
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "bgp_asn": null,
            "ec2_url": null,
            "ip_address": "2.3.4.5",
            "name": "AnsibleOffice",
            "profile": "shertel",
            "region": "us-east-1",
            "routing": "static",
            "security_token": null,
            "state": "present",
            "validate_certs": true
        }
    },
    "name": "AnsibleOffice"
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0
```

s-hertel committed:
Fix docs; gateway.customer_gateways is only sometimes returned, whereas gateway.customer_gateway always is. With this fix gateway.customer_gateways will not be returned at all since it provides no additional information.

Return the value of the Name tag instead of the metadata returned by tagging. This also allows results['name'] to be utilized when the customer_gateway already exists

Make the returned results concise by removing response metadata and customer gateways from the results

Add tags to the results returned during cgw creation

s-hertel committed:
fix RETURN documentation

willthames commented:
shipit

willthames commented on lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py:
```diff
@@ -255,6 +274,9 @@ def main():
                 )
+                created = gw_mgr.describe_gateways(module.params['ip_address'])
+                tag_array = created['CustomerGateways'][0]['Tags']
+                results['gateway']['CustomerGateway']['Tags'] = tag_array
```
Really need to run `boto3_tag_list_to_ansible_dict` on `tag_array`, but best to do that *after* `camel_dict_to_snake_dict` (otherwise tag keys get converted to lower case which can be unexpected)

s-hertel replied:
Good point! Addressed in clean_results(). The output is much nicer now (while also using #24897). Edit: #24897 has been merged so this is ready_for_review again.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py:
```diff
@@ -255,6 +274,9 @@ def main():
                 )
+                created = gw_mgr.describe_gateways(module.params['ip_address'])
+                tag_array = created['CustomerGateways'][0]['Tags']
+                results['gateway']['CustomerGateway']['Tags'] = tag_array
```
Ah. Yes!

s-hertel committed:
turn tags into an ansible dict (after making results snake case)

s-hertel committed:
Ec2_customer_gateway: don't delete the list of matching gateways; there can be multiple with the same IP, even if the module is not capable of creating them.

Don't modify tags with camel_dict_to_snake_dict

Add to RETURN docs

Always return the list of customer gateways

s-hertel committed:
Improve return docs

document returned tags are a dict

s-hertel commented:
ready_for_review

gundalow commented:
@s-hertel said we are OK to close, so closing
